Add MathProgBase solver interface.

tkoolen opened the pull request:
I know MathOptInterface is coming soon, but I want to try this with JuMP right now.

No tests yet, but most things seem to be working. Demo code:
```julia
using OSQP, JuMP

solver = OSQPSolver()
MathProgBase.setparameters!(solver; Silent = true)
m = Model(solver = solver)
@variable(m, x)
@variable(m, y)
@constraint(m, 1 <= x <= 4)
@constraint(m, y >= 3)
@objective(m, Min, x^2 + y^2)
status = solve(m)
@show getvalue(x)
@show getvalue(y)
@show getobjectivevalue(m)
```

For ease of implementation, I decided to store copies of `P`, `q`, etc. in `OSQPModel` and call `OSQP.setup!` in `MathProgBase.optimize!`. None of the updating infrastructure is currently used. Still, it's a start.

mlubin commented:
Looks reasonable. You should run the relevant functions in `MathProgBase/test/quadprog.jl` as part of the unit tests.

But I'd add that the important MOI infrastructure for QPs is now in place in JuMP, and we could use some early testers.

mlubin commented on src/OSQPSolverInterface.jl:
```diff
@@ -0,0 +1,186 @@
+    settings::Dict{Symbol,Any}
+end
+
+OSQPSolver() = OSQPSolver(Dict{Symbol, Any}())
```
Maybe name this `OSQPMPBSolver` to avoid the imminent conflict with MOI?

bstellato replied:
I see this is still the common approach in most of the current MathProgBase interfaces, e.g., [EcosSolverInterface.jl](https://github.com/JuliaOpt/ECOS.jl/blob/599d5bd365fe230110c66c92bd10c262b1532af2/src/ECOSSolverInterface.jl#L18). 

@mlubin Are you planning to change the other solvers as well or to just drop the MathProgBase interfaces altogether?

mlubin replied:
MPB interfaces will all be moved into separate packages when we start releasing the new MOI interfaces. It's up to developer discretion whether the MOI interface will sit in the same package as the direct solver wrapper or in a separate package.

bstellato replied:
Then it could be better to move this interface to a separate package called `MathProgBaseOSQP.jl` as done in [MathProgBaseMosek.jl](https://github.com/ulfworsoe/MathProgBaseMosek.jl). In this way we could keep `OSQPSolver` in both MPB and MOI interfaces and avoid using both them at the same time. Is this the purpose?

mlubin replied:
It seems a bit silly to create a package just for 200 lines of code that are going to be deprecated soon. (I'm reconsidering if we should do this for all solvers.) If we remove the line:
```julia
import .OSQPSolverInterface: OSQPSolver
```
then there's nothing much to worry about with having both MPB and MOI interfaces in the same package.

tkoolen replied:
Renamed the module to `OSQPMathProgBaseInterface` (solver is still named `OSQPSolver`). No longer exporting `OSQPSolver`, but exporting `OSQPMathProgBaseInterface` instead. What do you think?

bstellato commented:
Thank you for this. I guess the next steps are to:

- Complete the remaining functions that are compatible with OSQP
- Implement the problem data updates properly (this gives OSQP the real speedups by the way). I think this can be done by 
  * keeping track of the problem data as an internal copy of `P`, `q`, `A`, `l`, `u` 
  * calling the osqp function updates whenever possible, e.g., when `setobj!(m, c)` is called we call `OSQP.update!(model, q=c)`
  * having a boolean flag to define whether a new setup is required when `optimize!` is called. This is useful when funcitons like `addvar!` are called. 
- Add the quadprog tests as in [Gurobi.jl](https://github.com/JuliaOpt/Gurobi.jl/blob/master/test/mathprog.jl)

tkoolen commented:
> You should run the relevant functions in MathProgBase/test/quadprog.jl as part of the unit tests.

Since OSQP doesn't support quadratic constraints, neither `quadprogtest` nor `qpdualtest` is really applicable. There's also the additional issue of lack of support for variable bounds (do you agree with my handling of that by the way, or should they automatically be turned into constraints?). I opted to just copy over the relevant parts of `quadprogtest` for now. The dual-related functionality should still be tested somehow though.

mlubin commented on test/mpbinterface.jl:
```diff
@@ -0,0 +1,33 @@
+    sol = quadprog([0., 0., 0.],[2. 1. 0.; 1. 2. 1.; 0. 1. 2.],[1. 2. 3.; 1. 1. 0.],'>',[4., 1.],-Inf,Inf,solver)
+    @test sol.status == :Optimal
+    @test isapprox(sol.objval, 130/70, atol=1e-6)
+    @test isapprox(norm(sol.sol[1:3] - [0.5714285714285715,0.4285714285714285,0.8571428571428572]), 0.0, atol=1e-6)
```
I'm surprised the test passes with these tolerances.

mlubin replied:
Oops, missed the setting for eps_abs above.

tkoolen replied:
Actually, `polish = true` is more essential. Just setting `eps_abs` to even `1e-8` results in test failure.

mlubin replied:
Polish shouldn't be necessary, but you may need to set `eps_rel` (https://github.com/oxfordcontrol/osqp/issues/40) to a tiny value.

bstellato replied:
If you take the relative tolerance into account it might work without polishing:

```julia
@test isapprox(norm(sol.sol[1:3] - [0.5714285714285715,0.4285714285714285,0.8571428571428572]), rtol=1e-6, atol=1e-6)
```

mlubin commented on REQUIRE:
```diff
@@ -1,3 +1,4 @@
 BinDeps
 julia 0.6
 Compat
+MathProgBase 0.7
```
It's always a good idea to put an upper bound on the MPB requirement.

mlubin commented:
> There's also the additional issue of lack of support for variable bounds (do you agree with my handling of that by the way, or should they automatically be turned into constraints?).

Depends on whether you want to call OSQP on JuMP models that have variables with bounds.

bstellato commented:
> Since OSQP doesn't support quadratic constraints, neither `quadprogtest` nor `qpdualtest` is really applicable. There's also the additional issue of lack of support for variable bounds (do you agree with my handling of that by the way, or should they automatically be turned into constraints?). I opted to just copy over the relevant parts of `quadprogtest` for now. The dual-related functionality should still be tested somehow though.

I think people use variable bounds a lot. We could keep track of them in the problem data as
```
    lb <= Ax <= ub
    l <= x <= u
```

Then, when we call `optimize!` we stack them together. I can add it later today or tomorrow.

tkoolen commented:
Re: https://github.com/oxfordcontrol/OSQP.jl/pull/9#issuecomment-363144570: I started doing that initially, but I got a little worried about the interpretation of the dual variables becoming confusing/nonstandard.

We also need better handling of `sense`. For now, I've restricted `sense` to  be `:Min` in my latest commit in order to at least avoid blatantly wrong behavior. I'm not sure what the JuMP/MathProgBase convention is for handling duals when `sense == :Max`, so guidance would be appreciated. Also, the fact that `loadproblem!` contains the sense but not the 'P' matrix would have made subsequent calls to `setquadobj!` do the wrong thing for `sense == :Max`.

mlubin commented:
> For now, I've restricted sense to be :Min in my latest commit in order to at least avoid blatantly wrong behavior. I'm not sure what the JuMP/MathProgBase convention is for handling duals when sense == :Max, so guidance would be appreciated. 

JuMP has lots of tests for duals on LPs with both min/max sense. Not sure if there are specific tests for QPs. You should really feel free to only implement the parts that you need, because all of this is going to be tossed out after JuMP 0.18. Rejecting `Max` problems is reasonable.

tkoolen commented:
Improved test coverage. Also discovered and fixed a minor problem with `OSQP.warm_start!` along the way.

tkoolen commented:
Don't merge yet; I suspect there's something wrong with the quadratic objective methods.

bstellato commented:
Thanks for adding these tests. Tomorrow I will merge my changes as well to complete the interface.

tkoolen commented:
Fixed the objective bug and added yet more tests that displayed the problem. I'm successfully using OSQP through JuMP for my intended application now.

bstellato commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,197 @@
+    end
+end
+
+function Base.resize!(model::OSQPModel, n, m)
```
Is the `resize!` function really necessary? it seems that it does not properly resize the problem but it just creates zero matrices `P` and `A` and resizes the vectors. I guess this is needed in the `loadproblem!` function where you call `copy!`. Is this necessary/more efficient than just using `copy` or `deepcopy` for the data?

tkoolen replied:
FTR: answered on gitter. Yes, efficiency.

tkoolen commented on test/runtests.jl:
```diff
@@ -4,14 +4,15 @@ using Compat.Test, Compat.SparseArrays, Compat.LinearAlgebra, Compat.Random
+    # "primal_infeasibility.jl",
+    # "unconstrained.jl",
+    # "warm_start.jl",
+    # "update_matrices.jl",
```
Reenable

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+module OSQPMathProgBaseInterface
+
+using OSQP: Model, Results, setup!, solve!, update!, clean!, update_settings!, warm_start!
+importall MathProgBase.SolverInterface
```
`importall` is going away soon. See https://github.com/JuliaLang/julia/blob/master/NEWS.md#language-changes.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+    end
+end
+
+function get_qp_variables(model::OSQPMathProgModel)
```
Why not just have this be `numvar`?

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+    return n
+end
+
+get_qp_constraints(model::OSQPMathProgModel) = size(model.A, 1)
```
`numconstr`?

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+checksolved(model::OSQPMathProgModel) = isdefined(model, :results) || error("Model has not been solved.")
+
+# Reset problem when setup has to be performed again
+resetproblem(model::OSQPMathProgModel) = (model.results = nothing; model.perform_setup=true)
```
This shouldn't work. `model.results` can't store `nothing`.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+
+
+# Maybe add resetproblem function to remove results and reset perform_setup?
+checksolved(model::OSQPMathProgModel) = isdefined(model, :results) || error("Model has not been solved.")
```
This is no longer sufficient if problem modification is allowed. Actually, I think it was not correct before either.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+
+    # Update lower bounds in the model
+    if !model.perform_setup
+        update!(model.inner, l=[model.lb; model.l])
```
Allocation could be avoided, but not a huge deal.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+    
+    # Negate cost if necessary
+    if model.sense == :Max
+        model.q *= -1
```
Allocation could be avoided, but not a huge deal.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+getconstrmatrix(model::OSQPMathProgModel) = model.A
+
+
+function addvar!(model::OSQPMathProgModel, constridx, constrcoef, l, u, objcoef)
```
Lots of allocations, to the point that it may not be beneficial to support this. I don't think JuMP even uses this.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+    @boundscheck length(colidx) == nterms || error()
+
+    # Check if only the values have changed
+    if isdefined(model, :P)
```
This is always true.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+numquadconstr(model::OSQPMathProgModel) = 0
+
+# setquadobj!(model::OSQPMathProgModel, Q::Matrix) = ((Qi, Qj, Qx) = findnz(Q); setquadobj!(model, Qi, Qj, Qx))
+function setquadobj!(model::OSQPMathProgModel, rowidx::Vector, colidx::Vector, quadval::Vector)
```
This no longer fulfills the contract of `MathProgBase.setquadobj`.

> Duplicate index sets (i,j) are accepted and will be summed together

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+    # Check if only the values have changed
+    if isdefined(model, :P)
+        Pi, Pj, Px = findnz(model.P)
+        if (rowidx == Pi) & (colidx == Pj) & !model.perform_setup
```
Given that ordering of `rowidx` and `colidx` shouldn't matter and

> Duplicate index sets (i,j) are accepted and will be summed together

this is kind of a crude check.

bstellato replied:
I have left it for now since I cannot find an easier way to check if the sparsity pattern of P stays the same. If you have any suggestions feel free to update it.

tkoolen commented on src/mpbinterface.jl:
```diff
@@ -0,0 +1,506 @@
+
+    # Change sign if maximizing
+    if model.sense == :Max
+        model.P *= -1
```
Expensive way of doing this.

bstellato commented:
Damn, I pushed before finishing some stuff!!! Sorry. I had to rush out to finish some other things. I wanted to rebase everything before pushing it. I will adjust those things tomorrow ok? (Have lots of IKEA stuff to put together...). 

Thanks anyway for the performance tips. I am not totally familiar with Julia performance guidelines yet. I haven't used it that much until very recently.

bstellato commented:
I have fixed the interface. It should now support incremental model building, infeasibility certificates, Max or Min problems and problem data updates. There was a problem with the dual variables sign before. I have removed a couple of unnecessary allocations but I am quite sure we could do something better memory-wise. However, we might just leave it as it is and focus on performance improvements for the MathOptInterface. Let me know what you think.

tkoolen commented:
Looks good to me!
